state/protocol/protocol_state/epochs/identity_ejector.go

```diff
@@ -0,0 +1,68 @@
+package epochs
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+// trackedDynamicIdentityList is a helper structure for tracking identity lists in the state machine.
+// It is used to implement lazy initialization of the tracked identity list.
+// The structure relies on holding a reference to the list that is being modified.
+type trackedDynamicIdentityList struct {
+	dynamicIdentites flow.DynamicIdentityEntryList
+	identityLookup   map[flow.Identifier]*flow.DynamicIdentityEntry
+}
+
+// ejector is a dedicated structure for tracking ejected nodes in the state machine.
+// It is capable of tracking multiple identity lists and ejecting nodes from them.
+// The implementation is optimized for hot-path where ejections are rare, and by utilizing the lazy initialization,
+// the data structures are not populated until the first ejection is requested.
+// The ejector is used in the baseStateMachine to track ejected nodes and ensure that they are not readmitted during
+// the lifetime of the state machine.
+// It is not concurrency-safe.
+type ejector struct {
+	identityLists []trackedDynamicIdentityList
+	ejected       []flow.Identifier
+}
+
+// Eject marks the node as ejected in all tracked identity lists. If it's the first ejection during lifetime of the state machine,
+// the lookup tables are populated. The method returns true if the node was ejected in any of the tracked identity lists.
+func (e *ejector) Eject(nodeID flow.Identifier) bool {
+	l := len(e.identityLists)
+	if len(e.ejected) == 0 { // if this is the first ejection sealed in this block, we have to populate the lookup first
+		for i := 0; i < l; i++ {
+			e.identityLists[i].identityLookup = e.identityLists[i].dynamicIdentites.Lookup()
+		}
+	}
+
+	var ejected bool
+	for i := 0; i < l; i++ {
+		dynamicIdentity, found := e.identityLists[i].identityLookup[nodeID]
+		if found {
+			ejected = true
+			dynamicIdentity.Ejected = true
+			e.ejected = append(e.ejected, nodeID)
+		}
+	}
+	return ejected
+}
+
+// TrackDynamicIdentityList tracks a new DynamicIdentityList in the state machine.
+// It is not allowed to readmit nodes that were ejected. Whenever a new DynamicIdentityList is tracked,
+// we ensure that the ejection status of previously ejected nodes is not reverted.
```

Comment:
Thinking about this more, would it be worthwhile to include the epoch counter as a parameter to `TrackDynamicIdentityList`? That would communicate this requirement in a more obvious way than more comments, and enable the implementation to validate it is upheld.

```
TrackDynamicIdentityList(epochCounter uint64, list flow.DynamicIdentityEntryList) { ... }
```

We could hold one `uint64` in `ejector` and:
- set it when adding the first `list`
- check that `epochCounter == ejector.lastEpochCounter+1` when adding a new list, otherwise return an error

Reply:
I don't think we need to add them in epoch ascending order, it should work with any list composition.

Reply:
You're right, sorry for the confusion. 

I thought we were checking all ejected identities in `TrackDynamicIdentityList`, but we are only checking identities ejected in the current block.